### ISSUE TITLE
VCP alternate headwords, continue

### ISSUE THREAD

**Issue by funderburkjim** [sanskrit-lexicon/alternateheadwords#23]
This continues the work on Vachaspatyam alternate headwords, which is previously mentioned in #10, #11, and #17.

### readme1.md

The first step was to do a review of prior work. This review is in [readme1.md](https://github.com/sanskrit-lexicon/alternateheadwords/blob/master/data/VCP/readme1.md).

**Comment by funderburkjim** [comment]
### installation of ok1  list

Since the meta-line conversion (see [reference](https://github.com/sanskrit-lexicon/Cologne/issues/173) ) has been completed, we have a framework for installing alternate headwords for vcp.

The installation begins with the conversion of the [ok1](https://github.com/sanskrit-lexicon/alternateheadwords/blob/master/data/VCP/vcpahw2_hwnorm_ok1.txt) list. This is a list of 1066
alternate headwords out of the total 1740.  

For reasons described in the readme1 file, I think these are largely correct.

HOWEVER, they have not all been individually examined by hand.   

#### Help requested
If someone (I'm thinking of @SergeA) has the time or interest to do so, I can convert the ok1 list to Devanagari or IAST (whichever he prefers), so that he can review the correctness of the
algorithmically determined alternate headword spellings.   

It wold also be nice to have help in
resolving the 640 cases of the [todo1](https://github.com/sanskrit-lexicon/alternateheadwords/blob/master/data/VCP/vcpahw2_hwnorm_todo1.txt) list, so they may be
also installed as alternate headwords, and this task of alternate headwords for vcp may be completed.

**Comment by funderburkjim** [comment]
### ok1 list installed
These 1066 alternate headwords now installed at Cologne.

Next step is to work on the 640 cases of todo1 list as mentioned above.

**Comment by gasyoun** [comment]
> Since the meta-line conversion (see reference ) has been completed, we have a framework for installing alternate headwords for vcp.

Hurray, now we have a metalanguage for incorporating dictionary metastructure :100: 

> IAST (whichever he prefers)

Serge feels perfect with it, hope @SergeA will approve it.

**Comment by SergeA** [comment]
> If someone (I'm thinking of @SergeA) has the time or interest to do so, I can convert the ok1 list to Devanagari or IAST (whichever he prefers), so that he can review the correctness of the
> algorithmically determined alternate headword spellings.

I have time, but do not catch the task. Will be there an interface for this?
IAST is more preferable because it is easier to edit.

**Comment by funderburkjim** [comment]
### a possible UI
Marcis and I developed a UI for this task (see #17).  [Here](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?batch=test1) is a link.

Caveats:
* This link just covers a small test sample from the todo1 list.
* The link as shown uses SLP1, but by adding '&input=deva' to the url, Devanagari is used.
   *I should be able to modify so that adding '&input=iast' to the url will use IAST*

### Introduction to UI features
It is similar to the AP UI you worked with before, but not quite the same.  We were aiming the UI to
a mythical person named Radha, so there are some differences compared to the AP UI.
* The left-hand panel shows the cases.  
   * The first word for a case shows current spelling of the 'primary' headword
   * The second word shows a *suggested* alternate spelling
   * Click on a case to bring up the details in right-hand panel.
   * Controls at top (Next on Submit, etc.) provide alternate ways to navigate cases.
* The right-hand panel has 3 sections.
* The top section of right-hand panel
   * hw = `<primary headword>` 
   * shows suggested spelling of alternate headword as 'Old:'  (not editable)
   * repeats the suggested alternate headword spelling as 'New:'  --- **this is the field to edit**
      Initially the same as 'Old:'
   * Comments - self-explanatory; input optional
   * Submit button -- Click this when the 'New:' spelling is correct for the alternate spelling of the
     headword.   Corresponding case in left-hand-panel turns green to indicate case is done.
* The middle section of right-hand panel:
   * a list of dictionaries where the *primary* headword is found. 
   * clicking one of the dictionary buttons brings up the primary headword in that dictionary, in another tab
* The bottom section of right-hand panel:
  * a link to the scanned image in VCP -- the link tries to scroll near the word in question as an aid
     to finding the word on the page
  * The digitization of the entry
    * The most important part is on the line before the broken vertical bar `¦`. For instance,
       अनालोचि(डि)त¦  in case 2.  This is the 'raw' headword spelling of VCP. It determines the
        primary and alternate headword.
       * The primary headword is the part EXCEPT the part in parentheses.
        * The alternate spelling must be intepreted.  In case 2, it seems certain that the 
           correct interpretation is to replace the preceding चि with डि.  Dhaval's algorithm carried
           out this interpretation in order to arrive at the suggested spelling of the alternate headword
           that initially appears in the 'New:' field of the top panel.

### Main task of UI
 The primary goal is to get the correct alternate spelling of the headword.


### What list to use
My comment above was to have you check the 'OK1' list.   **HOWEVER**,  I think it makes more sense
to focus on the 'TODO1' list, and would suggest you start with that. 

@SergeA   If this looks like a good UI for the task and you want to proceed,  let me know, 
 I can add whatever spelling  convention you prefer  (such as IAST, or it could be HK if you would prefer that for the task).

**Comment by funderburkjim** [comment]
### Alternate to the UI
The UI tries to gather several bits of relevant information that might help in evaluating the correct alternate spelling of the headwords.

However, it is possible that a simple text file, similar to the [todo1.txt](https://github.com/sanskrit-lexicon/alternateheadwords/blob/master/data/VCP/vcpahw2_hwnorm_todo1.txt) file (but with IAST spelling) might be more convenient for this task.   It's really up to you which to use. Let me know.

**Comment by gasyoun** [comment]
>  I should be able to modify so that adding '&input=iast' to the url will use IAST

Please do so.

> a link to the scanned image in VCP -- the link tries to scroll near the word in question as an aid
to finding the word on the page

But without the green marking of the article as before?

> (but with IAST spelling)

Give him IAST, HK is not as easy, still far better than SLP1 for human reading.

**Comment by funderburkjim** [comment]
### iast available
[test1 iast](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?batch=test1&input=iast)  seems to properly encode and decode the test
batch in IAST.

Here's the full url:
http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?batch=test1&input=iast

**Comment by SergeA** [comment]
@funderburkjim 
Your explications are clear enough.
I'd prefer to do it through the UI.
The test link for IAST seems to be ok.
So I can start with what you'll give me.

**Comment by funderburkjim** [comment]
### First two batches
@SergeA  I'll separate the 'todo1' list into batches.  There are 670 cases.

I've made the first two batches of size 25.  Here are links to UI for these, using iast.

* [batch todo_01: cases 1-25](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?batch=todo_01&input=iast)
* [batch todo_02: cases 026-050](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?batch=todo_02&input=iast)

When these are done, I'll make up batches for the rest.   I can adjust the batch size (25) if you like.

### Links to rest of batches
[batch todo_03: cases 051-075](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_03)
[batch todo_04: cases 076-100](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_04)
[batch todo_05: cases 101-125](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_05)
[batch todo_06: cases 126-150](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_06)
[batch todo_07: cases 151-175](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_07)
[batch todo_08: cases 176-200](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_08)
[batch todo_09: cases 201-225](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_09)
[batch todo_10: cases 226-250](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_10)
[batch todo_11: cases 251-275](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_11)
[batch todo_12: cases 276-300](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_12)
[batch todo_13: cases 301-325](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_13)
[batch todo_14: cases 326-350](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_14)
[batch todo_15: cases 351-375](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_15)
[batch todo_16: cases 376-400](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_16)
[batch todo_17: cases 401-425](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_17)
[batch todo_18: cases 426-450](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_18)
[batch todo_19: cases 451-475](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_19)
[batch todo_20: cases 476-500](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_20)
[batch todo_21: cases 501-525](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_21)
[batch todo_22: cases 526-550](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_22)
[batch todo_23: cases 551-575](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_23)
[batch todo_24: cases 576-600](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_24)
[batch todo_25: cases 601-625](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_25)
[batch todo_26: cases 626-650](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_26)
[batch todo_27: cases 651-675](http://www.sanskrit-lexicon.uni-koeln.de/scans/VCPScan/2013/pywork/correctionwork/ahw-issue-017/update.php?input=iast&batch=todo_27)

**Comment by funderburkjim** [comment]
> But without the green marking of the article as before?

No, this is same as before.   
While the 'green' markup is always generated, it often doesn't show up - no idea why.
For me, the 'green' markup appears sometimes, and sometimes not -- usu. I have to reload the page once or twice for the green blob to show.  

Since the scrolling does most of the hard work of locating the word on the page, I usu. don't worry about the 'green' blob if it doesn't show up.

**Comment by funderburkjim** [comment]
## Progress Table for Batches

Since @SergeA is going to do them, I've prefilled all but the date field.


Batch | case1-case2 |  Date 
----- | ----------- | ---- 
todo_01 | 001-025 | 08/22/2017
todo_02 | 026-050 | 08/22/2017
todo_03 | 051-075 |
todo_04 | 076-100 |
todo_05 | 101-125 |
todo_06 | 126-150 |
todo_07 | 151-175 |
todo_08 | 176-200 |
todo_09 | 201-225 |
todo_10 | 226-250 |
todo_11 | 251-275 |
todo_12 | 276-300 |
todo_13 | 301-325 |
todo_14 | 326-350 |
todo_15 | 351-375 |
todo_16 | 376-400 |
todo_17 | 401-425 |
todo_18 | 426-450 |
todo_19 | 451-475 |
todo_20 | 476-500 |
todo_21 | 501-525 |
todo_22 | 526-550 |
todo_23 | 551-575 |
todo_24 | 576-600 |
todo_25 | 601-625 |
todo_26 | 626-650 |
todo_27 | 651-675 |

**Comment by funderburkjim** [comment]
@SergeA 

Let me know how it goes.  If the system looks right to you after the first two batches, I'll go ahead and
prepare all the batches for the todo list of 670 cases.

**THANKS FOR THE HELP !**

**Comment by gasyoun** [comment]
> While the 'green' markup is always generated, it often doesn't show up - no idea why.

Hmm, understood.

>  I have to reload the page once or twice for the green blob to show.

Same me.

**Comment by SergeA** [comment]
Batch todo_01 finished. But some problems with the second.

**Batch todo_02, case 8 "araṇya (ṇyānām)** " shows wrong article, though gives right link to the scan.
The headword must be corrected to "araṇya(ṇyānām)pati" (araṇyapati / araṇyānām pati).
I skipped it.

In **todo_02, case 10 there is अरिगूर्ण arigūrṇa.** Should we consider it as print error for अरिगूर्ण्ण arigūrṇṇa? I see VCP uses double consonants after R, but I don't know if it is systematic, or not. Chaotic spellings are troubling the search, of course.

**todo_02, case 14: aṣāḍhā(ḍ*ā)¦**
Do we use roman `ḍ*a` for the underdotted ड़ ?
Will it be processed correctly if I'll put asterisk in `"aṣāḍ*ā"`?

**todo_02, case 25: āśāḍh*a (āśāḍa)**
The entry shows error.
"key1entry problem:Error: jQuery310014695587162559132_1503403172096 was not called"

Is it right to write `ḍ*ha` for ढ़ ?

VCP spelling is inconsistent and freely interchanges ड&ड़, ढ&ढ़. It seems that authors did not differ these spellings.
In the case 14 it is given `aṣāḍhā(ढ) & aṣāḍ*ā(ड़)` and in 25 reversely `āśāḍh*a(ढ़) āśāḍa(ड)`. 
These two letters ड़ ढ़ represent the corrupted spelling, not classical. For the search and sorting they must be equal to ड ढ. Is it so?

**todo_02, case 15: asū(kṣa)kṣaṇa¦**
A bit not clear. I interpreted it as alternative word-endings asūkṣaṇa/ asūkṣa.

Also there was for a moment a problem with the connection to the database, the "submit" button did not effect, and then there was a data save error. But two-three minutes later it worked again good. Don't know if it is a problem of my connection or from the database side.

I'm ready to continue this way. 25 cases per batch are ok.

**Comment by funderburkjim** [comment]
### Responses to previous comments

*Batch todo_02 case 8*  Agree with your parsing of alternates as **araṇyapati / araṇyānām pati**.  Will have to handle slightly differently. Fine to leave it undone

In General, I'll skip normal processing for cases which are not done (as the 4 in batch todo_02).  These will have to be dealt with on individual basis.

*case 10*   VCP doubles the consonants after `r` pretty consistently.  I've put **arigūrṇṇa** as the alternate spelling, and submitted.   

> Incidentally -- We have developed a system of 'normalized spelling' (hwnorm1).  
   [Here](https://github.com/sanskrit-lexicon/Cologne/issues/171#issuecomment-323253168) are the
   current normalization rules.  hwnorm1 database comes into play in the UI.  For instance, in case 10,
   the MW button leads to headword 'arigUrta'  (one 't'), while the VCP button leads to 'arigUrtta' (two 't').

*cases 14 and 25:*  **ḍ** vs. **ḍ***    I think the `*` in the coding of VCP in slp1 represents 'nukta'; 
 and that asterisk passes through unchanged when converting to IAST. 
According to our digitization of vcp, it occurs 7200+ times, always after the soft cerebral consonants unaspirated or aspirated  (slp1 'q' or 'Q').  
From the Wikipedia article on [nukta](https://en.wikipedia.org/wiki/Nuqta):

```
Nuqtā (Hindi-Urdu नुक़्ता, نقطہ, from the Arabic nuqta نقطة "dot," or "period."), also spelled Nuktā, 
is a term for a diacritic mark introduced in Devanāgari (and some other Indian scripts) used to 
represent sounds from other languages which do not have a native character. 
It takes the form of a dot placed below a character. 
```

According to my recollection of a conversation with Peter Scharf several years ago, in regard to the use
of nukta in some other dictionary (I don't recall which dictionary),  there is no good reason to use it in
transcription of Devanagari Sanskrit.

Thus, I think we should retrochange all those 7200+ instances in vcp.txt (the change being to remove the `*` which represents nukta).

I think this agrees with @SergeA 's assessment:
>These two letters ड़ ढ़ represent the corrupted spelling, not classical. For the search and sorting they must be equal to ड ढ. Is it so?


I've submitted both cases 14, 25.   

**key1entry problem:Error:**  ~~~I'm not sure what causes this error message,   @SergeA  -- Let me know
when it occurs.   If the problem is sufficiently annoying, I'll try to find a solution.~~~

UPDATE:  the `*` is causing the problem when the dictionary data is being retrieved.  Problem solved, at least for case 25. 
Let me know if you see this message in any other cases in remaining batches.

*todo_02. Case 15*  asūkṣaṇa/ asūkṣa.   Agree with this solution.

**Comment by funderburkjim** [comment]
### Links for rest of todo cases:
The links for batches 3-27 have been inserted in comment [above](https://github.com/sanskrit-lexicon/alternateheadwords/issues/23#issuecomment-323904099).

**Comment by gasyoun** [comment]
> Thus, I think we should retrochange all those 7200+ instances in vcp.txt (the change being to remove the * which represents nukta).

Or make our search smarter, but leave the original encoding of letters.

**Comment by SergeA** [comment]
> Thus, I think we should retrochange all those 7200+ instances in vcp.txt (the change being to remove the `*` which represents nukta).

I do not see any good reason to keep these useless nuktas. These spellings do not show any word peculiarities, but only the accent of the authors of the dictionary. It is interesting. And if these nukta-spellings would be systematic, we could treat them as an orthographic nuance of the whole text (same as र्त vs. र्त्त) and to write corresponding rules for conversion. But as I can see, they are random and chaotic. 7200 asteriks are a big heap, of course... A heap of unnecessary troubles.

**Comment by SergeA** [comment]
> In General, I'll skip normal processing for cases which are not done (as the 4 in batch todo_02). These will have to be dealt with on individual basis.

I hope you've also noticed my comments as `agrediṣiṣu(ṣū)¦ >> agredidhiṣu(ṣū)¦` (todo_01, case 2) etc.

**Comment by drdhaval2785** [comment]
*>todo_02. Case 15* asūkṣaṇa/ asūkṣa. Agree with this solution.

I feel it should be asūrkṣaṇa as an alternate. Explanation and other dicts
support this

**Comment by SergeA** [comment]
> I feel it should be asūrkṣaṇa as an alternate. Explanation and other dicts support this

Yes, today I think this reading will be better.
VCP uses brackets in a spontaneous way, that leads to confusion. As this word is explained through sū(sū**r**kṣa)kṣa, which we should read as sūkṣa/sū**r**kṣa, it will be logically to suppose print error in the headword: asū(kṣa)kṣaṇa >> asū(**r**kṣa)kṣaṇa and read it as asūkṣaṇa / asū**r**kṣaṇa.
Thank you for correction.

**Comment by SergeA** [comment]
In batch 03 these ones give error:
22. ūrddha
23. ūrddhaka
24. ūrddhamanthin
Perhaps some problem with combination rddha/rddhva.

**Comment by SergeA** [comment]
**todo_05 Case 4** 
`kośa(ṣaṃ)¦` = kośa / koṣaṃ
But it must not be koṣaṃ, but koṣa, because in the headword it is given the base of the word. Am I right?

**todo_05 Case 11**
`kośā(ṣā)gāraṃ¦`
Once again there must not be ṃ here.
Should we say these to be print errors or what?

**Comment by SergeA** [comment]
**todo_05 Case 19**
`kau(ṣī)takī¦ `= kautakī / kauṣītakī ??
Looks like some error. As explained it is from ku**ṣī**taka, then wherefrom comes here kautakī? Perhaps some syllable is missing.

Also there is a typo here in explanations: kuṣītakasyāṣatyam >> kuṣītakasyāpatyam
(apatyam "offspring").

**todo_05 Case 25**
`kṣau(ṇa)ṇī¦` = kṣauṇī / kṣauṇa
But the word is marked strī, and kṣauṇa does not look like strī. Perhaps some error here.

After reviewing 125 words I'm already depressed by the poor quality of the text of VCP with so many print errors.

**Comment by SergeA** [comment]
**todo_06, Case 3**
`kha(rju)rjjūghna¦ pu0 kharjuṃ(rjūm)kaṇḍūyanaṃ`
once doubled rjj, trice not doubled rj
The real difference is in short/long vowel rju/rjū. And if doubling, there should be reduplication in both rjju/rjjū. Corrected this way. Am I right?
In the scan this place has print error: खट)र्जु)र्ज्जूघ्न.
 
The question of reduplication in VCP is not clear for me. Under the headword arjuna in VCP I count 5 **arjuna**s and 15 **arjjuna**s. IMHO the authors liked to write rjj, but did not care too much if it was written as rj. How should we treat this situation? Have we to systematize the orthography? Should we correct the headword arjuna to arjjuna?

**Comment by SergeA** [comment]
**todo_06, Case 8**
`gandhakheḍa(la)(ka)¦ `
gandhakheḍa / gandhakhela / gandhakheḍaka / gandhakhelaka 
 4 variants!

`+`
**todo_08, Case 23**
`dāśa(sa)pu(pū)ra¦`
dāśapura / dāsapura / dāśapūra / dāsapūra

`+`
**todo_10, Case 25**
`paśa(ṣa—sa)¦` & `paśa(ṣa)(sa)`
paśa / paṣa / pasa
3 variants

**Comment by SergeA** [comment]
**todo_10, Case 7**
`nāgavi(bila)¦ na0 tīrthabhede nāgarakhaṇḍaśabde dṛśyam .`
Not clear.
Perhaps it is supposed nāga-bila "snake's hole" with variative b/v nāga-vila. Or perhaps it means a word nāgavi in the sense of nāgabila? The given link "tīrthabhede nāgarakhaṇḍaśabde" points, as I understand it, to Skanda-purana, but I did not find its digitalized version for search.

**todo_08, Case 5**
`tumba(ṣa)vana¦ `
I interprete as tumbavana / tuṣavana, though it seems suspicious. MW has only tumba-vana  "N. of a place", so -vana should not be changed to ṣana, so remains only tumba>tuṣa.

**Comment by gasyoun** [comment]
> After reviewing 125 words I'm already depressed by the poor quality of the text of VCP with so many print errors.

Print or digital errors?

> IMHO the authors liked to write rjj, but did not care too much if it was written as rj. 

Agree totally.

> Should we correct the headword arjuna to arjjuna?

I would go this way, @drdhaval2785 do you support it?

**Comment by SergeA** [comment]
> >  poor quality of the text of VCP with so many print errors.
>
> Print or digital errors?

Print errors do mean print errors. For an accurate text this task could be finished in one day. But after every few words I take a break for meditating on lost, corrupted and misplaced syllables or on some irregularities in the authors' orthography etc. Authors' inconsistencies are annoying as well.

**Comment by SergeA** [comment]
Finished all batches except those cases which
1) give error
2) show multiple articles
3) have more than 1 alternative reading
4) seem questionable

**Comment by SergeA** [comment]
**todo_26, case 21**
velli (vellī)
Shows 2 articles. Skipped.

**todo_27, case 6**
sāma (ṣāma)
Shows 2 articles. Skipped.

**Comment by SergeA** [comment]
**todo_12, case 16**
pūrva(rba)bha(bhā)drapadā¦
pūrvabhadrapadā / pūrvabhādrapadā / pūrbabhadrapadā / pūrbabhādrapadā
4 alternatives

**todo_13, case 14**
pūrvā(rbā)hṇa(hṇe)tana¦
4 alternatives

**todo_14, case 5**
paurvārddha(rddhika)rddhya¦
Here there must be a print error. There should be 3 alternatives:
paurvārddha / paurvārddhika / paurvārddhya
Skipped.

**todo_15, case 4**
ba(va)hutvac(kka)¦
bahutvac / vahutvac / bahutvakka / vahutvakka

**todo_15, case 10**
ba(va)hupāda(d)¦
bahupāda / vahupāda / bahupād / vahupād

**todo_16, case 14**
ba(va)hvṛc(ca)¦
bahvṛc / vahvṛc / bahvṛca / vahvṛca

**todo_18, case 23**
bra(vra)hmajā(yā)mala¦
brahmajāmala / vrahmajāmala / brahmayāmala / vrahmayāmala

**todo_23, case 22**
bra(vra)hmoḍu(du)mbara¦
4 alternatives

**~~~todo_23, case 22~~~**
**todo_25, case 6**
bhavādṛkṣa(śa)ś¦
bhavādṛkṣa / bhavādṛśa / bhavādṛś

**todo_25, case 21**
yādṛkṣa(śa)(ś)¦
yādṛkṣa / yādṛśa / yādṛś

**todo_26, case 4**
laguḍ*a(la)(ra)¦
laguḍa / lagula / lagura

**todo_26, case 19**
vṛ(bṛ)ṣi(ṣī)¦
vṛṣi / bṛṣi / vṛṣī / bṛṣī

**todo_26, case 24**
śarāṭi(ḍiṃ)(li)¦ >> śarāṭi(ḍi)(li)¦
śarāṭi / śarāḍi / śarāli

**Comment by SergeA** [comment]
Usually cases with double brackets give 3-4 alternatives. But there are cases where the second brackets are depending on the first.

**todo_12, case 13**
pūrva(rba)pūrva(rba)¦ = pūrvapūrva / pūrbapūrba
Technically there can be 4 alternatives here, but from the point of common sense I think there can be only 2, because it is repetition of the same word.

**todo_14, case 19**
baṣka(ska)yi(ṇī)nī¦ >> baṣka(ska)yi(nī)ṇī¦
Here 2 alternatives: baṣkayiṇī / baskayinī, where the spelling n/ṇ depends on previous s/ṣ.
The authors made an error with brackets in this headword (but in explanations did it correctly).

**Comment by SergeA** [comment]
It is not unfrequent when the alternatives are not spelling variations but represent different words with some common part. E.g. 
dhūstarā(mā) = dhūstarā / dhūstamā
Here words with different suffixes tara/tama.

Another funny example:
bhogīndra(śa) = bhogīndra / bhogīśa
Here indra and īśa are represented as alternative readings.

And for the last example is the last word of the list (todo_27, case 20):
hiraṇyavāha(hu)¦
Complicated case where different words vāha "bearer" and bāhu "arm" are combined by indiscrimination of b/v. So the spelling of the alternatives is hiraṇya-vāha "gold-bearer" and hiraṇya-bāhu "golden-armed". 
I propose to change this headword as: 
hiraṇyavāha(bāhu)¦

**Comment by SergeA** [comment]
I suppose all the headwords in VCP ending in M or Ṃ and without the mark "avyaya" are typos and print errors. **M**s म often come from badly printed **S**s स, and **Ṃ**s ं are just blobs.

**Comment by SergeA** [comment]
The situation with the orthographic consistency in VCP is sad. The letters ब व are treated as equivalent though graphically different. Herefrom come stupid long lists with pūrva/pūrba, bahu/vahu, bṛhat/vṛhat, brahma/vrahma etc. Even if alternative readings are not pointed in the article, they are always meant and can come up at any place. E.g. the word vīra is given without bīra variant, but for mahāvīra is given also mahābīra. 
And about situation with consonant duplicating after repha I've already said. As if people that composed VCP were in continuous controversies on how to write words.  Or rather they didn't care too much about following any strict system.

**Comment by SergeA** [comment]
**todo_07, case 6**
caturṇṇa(na)vati¦ = caturṇṇavati / caturnavati
Not sure if will be good here caturnavati or caturnnavati with duplication.

**Comment by SergeA** [comment]
BTW, I have no rights to edit the table of batch-case-date.

**Comment by funderburkjim** [comment]
Re todo_02, case 15.  Changed UI solution to asūrkṣaṇa 
in light of above [comment](https://github.com/sanskrit-lexicon/alternateheadwords/issues/23#issuecomment-324288472)

**Comment by funderburkjim** [comment]
> In todo_02, case 10 there is अरिगूर्ण arigūrṇa. ... or arigūrṇṇa (and arigūrṇa a print error).

In body of this entry,  `arigūrṇṇa` is shown.  So text is internally inconsistent.  No way to conclude which is 'right'.

Incidentally, I found myself confused as to how `ṇṇ` and `ṇ` differ in the Devanagari of VCP.
I think the difference is that `ṇṇ` has a 'bar' between the 3 vertical lines, and these are absent in `ṇ`:
![image](https://user-images.githubusercontent.com/6393033/29844253-9c385484-8cdc-11e7-88ad-d1f06754d242.png)

ADDENDUM:
 In todo_01, case 22., see `abhigūrtta, abhigūrṇṇa` --  This evidence supports contention that `arigūrṇa` is a print error.

**Comment by funderburkjim** [comment]
> todo_02, case 14: aṣāḍhā(ḍ*ā)¦

Best to leave out the '*', as your current solution does.   However, I will examine all the '*' cases, and, as
mentioned elsewhere, will remove all these 'nukta' codings in the dictionary as a whole.

**Comment by funderburkjim** [comment]
> I hope you've also noticed my comments 

Thanks for reminder!  I've now worked through all the todo_01, todo_02 UI cases.
For several cases (notably the aBra/abBra compounds), I've added comment that MW spelling confirms the `aBra` spellings.

**Comment by SergeA** [comment]
> I think the difference is that ṇṇ has a 'bar' between the 3 vertical lines, and these are absent in ṇ

Yes, this ligature is written this way.

> So text is internally inconsistent. No way to conclude which is 'right'.

We can see a common tendency to duplication after repha. So I suppose the duplicated variant fits better. But this inconsistency is a real problem in the whole VCP, and if we want to unify the text's orthography, we have to make numerous changes.

**Comment by SergeA** [comment]
> For several cases (notably the aBra/abBra compounds), I've added comment that MW spelling confirms the aBra spellings.

MW gives also अब्भ्र variant in the अभ्र article.

**Comment by funderburkjim** [comment]
> todo_05 Case 4
kośa(ṣaṃ)¦ = kośa / koṣaṃ
But it must not be koṣaṃ, but koṣa, because in the headword it is given the base of the word. Am I right?
Case 11, similar

Yes, I think that VCP gives the base for substantives.   Thus, we should consider the `ṃ` in `koṣaṃ` to be a print error.

**Comment by funderburkjim** [comment]
> todo_05 Case 19
kau(ṣī)takī¦= kautakī / kauṣītakī ??

For this one, I'll treat as `kauṣītakī`,  with NO alternate.   Will consider the parentheses in `(ṣī)` to be a print error.

**Comment by funderburkjim** [comment]
> once doubled rjj, trice not doubled rj

Here's a way to think of this inconsistency.   

I think there is a sandhi rule which says that a consonant after 'r' can be *optionally* doubled.  @drdhaval2785 has probably quoted to me the exact sandhi rule for this, but I can't locate it now.

Assuming this rule,  then *either* spelling  (doubled or not) is grammatically correct.  

We might expect that a given author would make a *global* choice of one of the two options, and consistently use that option within a given work.  However,  there is nothing in the grammar that mandates such consistency.

So we can be annoyed that there are optional sandhi rules in Sanskrit  (I certainly find it annoying).
But we can't say that an author is grammatically wrong because he (in effect) flips a coin each time he encounters a situation where an optional sandhi rule is applicable, and uses the coin toss result as his guide to which option to use this time.

**Comment by funderburkjim** [comment]
> Should we correct the headword arjjuna to arjuna ?

I don't think we need to correct it, but we need to make provision for this in our search algorithms.
This is where the notion of 'normalized spelling' comes in.   So far, this normalized spelling is not
widely used in the Cologne displays.   However, it does come into play in the recent 'simple input' scheme used in the [list-0.2s](http://sanskrit-lexicon.uni-koeln.de/scans/awork/apidev/simple-search/v1.0/list-0.2s.html)  display  using the 'simple' input option.    I'd be interested to know if you find that this display handles the doubling or non-doubling after 'r' in a helpful way.

**Comment by funderburkjim** [comment]
### todo_07, case 6

caturṇṇa(na)vati   caturṇṇavati / caturṇnavati

I think the alternate should be caturnavati.    Reasons:
* Note `caturṇa(na)vata ` in entry.   Which suggests the retroflex `ṇ` is being replaced by dental `n`.
* MW has caturnavati.

@SergeA  Agree?

**Comment by funderburkjim** [comment]
Since there are so many cases that need special attention,  I'm going thru the UI and adding a comment 'TODO' or prefixing a comment similarly if there is already a comment.  In the installation process, I'll partition off the cases so marked for a second round of special handling.  This will allow the bulk of the cases, which have no 'TODO' comments (either in UI or in this issue) to be handled in a 'normal' way.  Then, the smaller set of special handling cases will examined individually.
Currently this marking is thru batch todo_08.

**Comment by SergeA** [comment]
> I think there is a sandhi rule which says that a consonant after 'r' can be optionally doubled. 

> Assuming this rule, then either spelling (doubled or not) is grammatically correct.

>  ... flips a coin each time he encounters a situation where an optional sandhi rule is applicable ...

`अचो रहाभ्यां द्वे। Pāṇini viii.4.46`

Vasu tr.: 
`There is reduplication of यर् , i.e. all the consonants except ह , after the letters  र् and ह् following a vowel. The word यर् of the last sūtra is understood here. According to others, the वा is also understood, and this is an optional rule. Thus अर्क्कः , मर्क्कः , ब्रह्म्मा , अपह्न्नुते ॥`

Actually I've never seen reduplications after ह like **brahmma** etc. (Except some editions of Yajurveda, where a huge heap of additional reduplication rules were observed, and every other consonant resulted reduplicated.) Large number of manuscripts and books support reduplication after र. The optional rule of reduplication represents two orthographic traditions (which is connected with pronunciation of course). Usually the author or copyist belongs to one or another tradition and makes efforts to keep the text within the tradition. So in one manuscript we see अर्जुन and in another one अर्ज्जुन. Perhaps with rare erroneous deviations. But they should not continuously interchange! I'll give an example from Russian. The Russian word "god" when it is written by christians and other religious men take first capital **Бог** "God", as a sign of respect. But atheists in the same context will write it with small letter **бог** "god". Both spellings are correct. But it is impossible to imagine somebody's flipping a coin every time to check if he believe in god or not and deciding how to write God or god. The same way I cann't imagine a pandit flipping a coin every time to ascertain if he follows the optional sandhi rule or not, except he's an absolute schizo!

**Comment by SergeA** [comment]
> >    todo_05 Case 19
> >     kau(ṣī)takī¦= kautakī / kauṣītakī ??
> 
> For this one, I'll treat as kauṣītakī, with NO alternate. Will consider the parentheses in (ṣī) to be a print error.

I also incline to this decision.

> 

>> todo_07, case 6
>> caturṇṇa(na)vati caturṇṇavati / caturṇnavati
> 
> I think the alternate should be caturnavati. Reasons:
>     Note caturṇa(na)vata in entry. Which suggests the retroflex ṇ is being replaced by dental n.
>     MW has caturnavati.
> @SergeA Agree?

Here two moments:
1) ṇatva after R - rna/rṇa - this is the matter of the alternate reading here
2) duplication after R - here again comes to scene author´s inconsistency

Logically it should be symmetric: rna/rṇa (as in MW) or rnna/rṇṇa (in the case of unified reduplication).
If we accept the VCP's chaotic orthography of duplications, then any spelling is acceptable, and there is no problem with this case, only to correct the alternate reading after printed: 
catu**rṇṇ**avati / catu**rn**avati

Or perhaps here the logic must be to consider this word as one or as compound of two and then in a single word we get ṇatva and reduplication (>rṇṇ), but with no sandhi between compounded two words (>r-n). I am not smart enough in these grammatical fine moments...

**Comment by drdhaval2785** [comment]
> caturnavati

Agree

**Comment by funderburkjim** [comment]
### todo_14, case 14

ba(va)daraballī    I think typo  (print hard to read) >> ...vallī .   Although b/v often interchanged,  VCP has headword `vallī` but not `ballī`

**Comment by funderburkjim** [comment]
### todo_24, case 4 and 5
The print shows alternates `brā(vra)`  in both cases.  Should the alternate also have ā ?
So suggest print errors:
* brā(vra)hmaka¦  >> brā(vrā)hmaka¦
* brā(vra)hmagupta¦ >> brā(vrā)hmagupta¦ 

@SergeA Agree?

**Comment by SergeA** [comment]
> ba(va)daraballī I think typo (print hard to read) >> ...vallī 
> brā(vra)hmaka¦ >> brā(vrā)hmaka¦
> brā(vra)hmagupta¦ >> brā(vrā)hmagupta¦

Agree.

**Comment by SergeA** [comment]
> the list-0.2s display using the 'simple' input option. I'd be interested to know if you find that this display handles the doubling or non-doubling after 'r' in a helpful way.

Yes, I think this new feature is very useful and great improvement. The user can get Arjuna without worrying about his spelling in VCP with rj or rjj.
Though when I requested in VCP the word **ardha** it didn't find it, nor the exact VCP spelling **arddha**. Some bug.

**Comment by funderburkjim** [comment]
### installation begun

Preliminary phase of installation (parsing the results from UI) is done.

Of the 670 cases, 581 are done and 89 remain to be done.
    These 89 are the ones marked in UI comment as 'TODO', for one reason or another.

As an aside,
Of the 581 done cases
    452 had no change in spelling of alt key (i.e., Dhaval's agorithmic deduction confirmed by Serge)
   and 129 had changes.

Tomorrow I'll deal with:
* removing the nukta coding
* handling the 89 special cases, including correcting print and typo errors Serge noted
* finishing the installation of all the 670 alternate key cases from batches todo_01 through todo_27.

**Comment by funderburkjim** [comment]
> ardha it didn't find it, nor the exact VCP spelling arddha

I'm sure this is because the 'doubling' has two different letters after the 'r', unaspirated 'd' and aspirated 'dh'.

First, I'm not sure why neither spelling was found. - Will need to investigate, surely a bug.

But also, this will require a new rule, probably in the spelling normalization function.

How do you think such a rule should be stated?   Maybe,  'r' + X + AX -> 'r' + AX ?   (where X is
an unaspirated consonant and AX is the corresponding aspirated consonant).

Should such a rule also hold where instead of 'r' we have vocalic r `ṛ`  (what about long vocalic r `ṝ`?)

**Comment by funderburkjim** [comment]
### todo_03: 22-24
22. ūrddha
23. ūrddhaka
24. ūrddhamanthin
The problem (non-retrieval of VCP data in UI) was due a prior incomplete correction.

The printed entry (e.g. for 23. ūrddhaka) is  (in SLP1 notation):
```
UrdDa(dDva)ka¦ pu0 UrdDaH san kAyati SabdAyate . mfdaNgaBede sa
```
This implies a 'key2' of `UrdDa(dDva)ka`.  The *normal* key1 implied by this key2 is `UrdDaka`.
However, since sometime it was noticed that the *usual* spelling is `UrdDvaka`, it was decided to
make key1 to be UrdDvaka.  So when the UI tried to retrieve information under the implied key1 `UrdDaka`, none was found.

Since we now have the luxury of alternate headwords,  the earlier change is not needed.
So `UrdDaka` will be the primary headword spelling, and `UrdDvaka` the alternate.

Similarly for cases 22 and 24.

ADDENDUM:
This could easily be viewed as a different kind of inconsistency in VCP.  The reason is that MANY
additional headwords are spelled (without alternates) with prefix 'UrdDva', but only the three above
plus also 4 more (UrdDa(rdDva)kaca,  UrdDa(rdDva)kaRWa, UrdDa(rdDva)karmman, UrdDa(rdDva)mAna)
are shown with prefix `UrdDa`, and there are NO examples of headwords with prefix `UrdDa` which do not also show the `UrdDva` form as an alternate.  So, it would have been more logical to show these 7 with primary form `UrdDva` and secondary form as `UrdDa`.

**Comment by funderburkjim** [comment]
### todo_06, Case 3, additional observation

kha(rju)rjjūghna¦ >> kha(rjju)rjjūghna

Interestingly, among the headwords of vcp, as printed, we see that the 'j' is not doubled before the short 'u', but is doubled before the long 'u'; i.e. `kharju`  and `kharjjū` are headwords.

**Comment by funderburkjim** [comment]
> BTW, I have no rights to edit the table of batch-case-date.

@gasyoun  Do you know how to give @SergeA the rights?

**Comment by funderburkjim** [comment]
### Re. todo_10, Case 7
nāgavi(bila)    In SKD, There is `vi(bi)laṃ¦, klī, (vila + kaḥ .) chidram . ityamaraḥ `.

So in VCP, looks like print error in placement of parentheses 
nāgavi(bila)  >> nāgavi(bi)la .  We've seen this kind of paren error elsewhere in VCP, so going with
this interpretation.

**Comment by funderburkjim** [comment]
### todo_10, case 14

neulī / neḍulī

 I can find no whiff of either in MW.  Any idea of an English rendering of the meaning?

**Comment by drdhaval2785** [comment]
@SergeA has been given collaborator righy now. So he can edit posts now.
I think ideal would be to include him as team member of organization of sanskrit lexicon.

**Comment by SergeA** [comment]
> neulī / neḍulī
> I can find no whiff of either in MW. Any idea of an English rendering of the meaning?

This yogic practice is more known as naulī.
https://en.wikipedia.org/wiki/Nauli

**Comment by gasyoun** [comment]
> I think ideal would be to include him as team member of organization of sanskrit lexicon.

Done. Changed @SergeA from Member to Owner.

> I think the difference is that ṇṇ has a 'bar' between the 3 vertical lines, and these are absent in ṇ:

Yes, and not only in VCP, actually.

>  if we want to unify the text's orthography, we have to make numerous changes.

I would go for it.

> But we can't say that an author is grammatically wrong because he (in effect) flips a coin each time he encounters a situation where an optional sandhi rule is applicable, and uses the coin toss result as his guide to which option to use this time.

Even if he remains right for our purpose of quick search - we want to have it one way. With doubling or without, but one and only one.

> This could easily be viewed as a different kind of inconsistency in VCP

Dhaval, maybe add it to the dictionary standard page?

**Comment by funderburkjim** [comment]
> nauli -

I've heard of nauli.  Why does VCP spell it 'n**eu**li'  ?

**Comment by SergeA** [comment]
> Should such a rule also hold where instead of 'r' we have vocalic r ṛ (what about long vocalic r ṝ?)

No. This duplication takes place only after consonantal R.

> How do you think such a rule should be stated? Maybe, 'r' + X + AX -> 'r' + AX ? (where X is
> an unaspirated consonant and AX is the corresponding aspirated consonant).

Unaspirated or aspirated consonants after R both are doubling by inserting an unaspirated consonant in between, e.g. r+d = r**d**d ; r + dh = r**d**dh. So for normalization we have to do it backwards.
'r' + X + X -> 'r' + X
'r' + X + AX -> 'r' + AX 

However there are certain situations where similar double consonant cases are the result of combination of the root with the suffix.
vārddha < vṛddha < vṛdh+ta
vārtta < vṛtta < vṛt+ta
karttṛ < kṛt + tṛ
karttavya < kṛt + tavya
So here is the danger to process such cases as if duplicated after repha and then to get wrong forms.
kartṛ 'one who does' (from the root kṛ) != karttṛ 'one who cuts' (from the root kṛt)

**Comment by SergeA** [comment]
> Why does VCP spell it 'neuli' ?

VCP quotes from Rudrayamala, where the term is spelt this way - **neulī**. If VCP be quotting Hathayogapradipika the spelling would be **nauli**. And in Gheranda-samhita it is **laulikī**. Who knows why...

**Comment by funderburkjim** [comment]
### alternate Headwords installed

All these todo1 alternate headwords are installed,  including the numerous corrections for print changes and typos.

#### One exception: todo_05, case 25
The only exception (AFAIK) is `kṣau(ṇa)ṇī`,   where no alternate thus far added. (see [comment above](https://github.com/sanskrit-lexicon/alternateheadwords/issues/23#issuecomment-324796597)).


### sanhw1 diff is interesting
It is interesting to see the effect of these changes in the sanhw1 file of the CORRECTIONS repository
[link to history](https://github.com/sanskrit-lexicon/CORRECTIONS/commit/00b30f1e3b058e943bd73de5fbb4134512015393#diff-fb58e89c06d3f8195f8d0d718b181dfd).

**Comment by funderburkjim** [comment]
> Unaspirated or aspirated consonants after R both are doubling by insertion an unaspirated consonant in between, e.g. r+d = rdd ; r + dh = rddh. So for normalization we have to do it backwards.
'r' + X + X -> 'r' + X
'r' + X + AX -> 'r' + AX

Agree.   I think this is what I intended to say.  Incidentally, the logic of the normalization is applied when
all spellings are in SLP1.  Since the unaspirated and aspirated letters are represented by distinct *single*
letters in SLP1  (e.g., d and D),  there is no confusion:  rdd -> rd, and rdD -> rD are the normalizations.

**Comment by funderburkjim** [comment]
> Normalization problems with vftta, etc.

I've been aware of an undesirable mingling of the normalizations of 'vftta' and 'vfta'.  Here is the
current hwnorm1c record:
```
ALL OF THESE ARE IDENTIFIED  (SLP1 spelling)
vfta:  from root vf, to cover
  vfta:AP,AP90,BUR,CAE,CCS,MD,MW,MW72,PW,PWG,SHS,SKD,STC,VCP,WIL,YAT;
  vftaH:SKD;
 vftta: from root vft, to turn
  vftta:AP,AP90,BOP,BUR,CAE,CCS,INM,MCI,MD,MW,MW72,PE,PUI,PW,PWG,SCH,SHS,STC,VCP,WIL,YAT;
  vfttaM:SKD;vfttaH:SKD
```

And you have brought up several additional cases like karttf/kartf  (SLP1 spelling).

From the point of view of commingling words only when they are mere spelling variants (not semantic variants),  the current normalization system is at fault in such cases.

However, from the point of view of search utility, maybe this commingling is useful.  

> This duplication takes place only after consonantal R

If this is right, then  my 'vftta/vfta'  example would be solved. 
Reason:  the current normalization rule includes (in addition to consonant 'r') vocalic r:
```
'f' + X + X -> 'f' + X  ('f' is slp1 spelling of vocalic 'r' = IAST `ṛ`).
```

And by your quoted `अचो रहाभ्यां द्वे। Pāṇini viii.4.46`  rule above,  only consonantal 'r' is mentioned.
So **'f' + X + X -> 'f' + X' is wrong** !
So  I should remove it.

Will do so soon.



Good! Progress.

**Comment by drdhaval2785** [comment]
Good progress.
First sizeable incorporation of alternate headwords in other than MW dict.

**Comment by gasyoun** [comment]
> However there are certain situations where similar double consonant cases are the result of combination of the root with the suffix.

So work for some whitelisting.

Thanks, @SergeA

**Comment by funderburkjim** [comment]
#### Removal of 'fxx' -> 'fx' rule in hwnorm1

See discussion above for reason.
See the `hwnorm1c.txt` part of [this diff](https://github.com/sanskrit-lexicon/hwnorm1/commit/45bc463e95f0211cfebb5e3642001dc9af485b80#diff-56749a42336db113a1a94610525b3eaf)  which shows the impact of this rule change on the list of normalized spellings.

**Comment by funderburkjim** [comment]
#### Add  'rxX' -> 'rX' rule in hwnorm1
where X is aspiration of x.
e.g. ardDa -> arDa.  See discussion above.

See `hwnorm1c.txt` part of [this diff](https://github.com/sanskrit-lexicon/hwnorm1/commit/0d37d84c57086ccddf50be62cacdec07090e0c2b#diff-d00984d1651c7ea7c089fe25fd17b0f5) for details.  Note you will have to 'load diff' to see the instances.

**Comment by funderburkjim** [comment]
#### No whitelisting 
The normalization process includes no whitelisting thus far.
The normalization rules can be seen by looking at the ' normalize_key' function in the Python program
[hwnorm1.py](https://github.com/sanskrit-lexicon/hwnorm1/blob/master/ejf/hwnorm1c/hwnorm1c.py).

The comments (lines starting with `#`) should make the rules understandable by non-programmers.

@SergeA  and @drdhaval2785 probably should examine these rules:
*  are there any  erroneous rules ?
*  are there any missing normalization rules ?
* are there any rule exceptions for which we should develop a whitelisting component to the normalization ?

**Comment by SergeA** [comment]
`#4. ending 'aH' is 'a' (Apte)`
This must not be applied to adverbs (marked 'adv.'). There is an inconsistency in Apte 90 spellings. Some adverbs are written through -s (kutas, tatas, bhūyas, madhyatas &c.), but some through visarga (akutaḥ, akāmataḥ, kāmataḥ &c.). Note that with the same base kutas, but a-kutaḥ. These -aḥ adverbs better be normalized to -as.

`#7. 'cC' is 'C'`
What is the purpose of this rule? This certainly will corrupt many spellings like gacchat, acchāya, ucchid, ucchvas &c.

**Comment by funderburkjim** [comment]
> ending aH is a  not right for adverbs

This refinement is not possible with the current implementation of the normalization protocol.  The 'normalization function' is a function `f` that takes a given spelling X and returns a normalized form Y of that spelling.  

We can imagine a better normalization function `g` which might have *two* inputs (X,G) where G is
the some coding of grammatical information; and then g(X,G) would depend not only on the spelling X
but also on the grammatical form G.    

To apply `g` to all the entries of a given dictionary we would first have to have `G` for those entries.
This lexical or grammatical informatiion `G` is currently available only for MW.

There are two kinds of uses that I see for this normalization; they are loosely related but not identical.
* aligning entries in dictionaries
   We know that for two dictionaries D1 and D2, many of the entries are comparable.   
   For entry E1 in dictionary D1 and entry E2 in dictionary D2, we would like to know  if E1 and E2 are
   about the 'same word`.  Theoretically, there is probably a function EQ such that
   EQ(D1,E1,D2,E2) is true or false according to whether the two entries are about the same word or not.
  The normalization function `f` as currently implemented can be used to define a first approximation
  to EQ, using  a formula
  `EQ(D1,E1,D2,E2) = true if f(X1) is same spelling as f(X2), or  = false otherwise`
  Where X1 and X2 are the citation forms of the entries E1 and E2.
* for searching among several dictionaries
  Suppose a user enters a query X and we want to find all the entries in all the dictionaries with X as
  headword.  If X is `vanam`, then the user probably wants to find the entry in AP90 with headword `vanaM` and the entrie in AP with headword `vanam` and the entry in MW with headword `vana`, etc.  One
  way to accomplish this is to:
  * normalize the user's query: f('vanam') = 'vana'
  * precompute normalizations of ALL headwords in all dictionaries.  That is what [hwnorm1c.txt](https://github.com/sanskrit-lexicon/hwnorm1/blob/master/ejf/hwnorm1c/hwnorm1c.txt) contains.
  * Look up 'vana' in the precomputed list.  The current entry tells us that there are two spellings
    whose normalization is `vana`:
    * `vana`  is headword in many dictionaries: BEN, BHS, BOP, BUR, CAE, GRA, IEG, KRM, MD, MW, MW72, PGN, PUI, PW, PWG, SCH, SHS, SKD, STC, VCP, VEI, WIL, YAT
    * `vanaM` in AP90, SKD
    * NOTE:  we don't find `vanam` (so we miss the entry in AP90). Maybe we should add rule `am` -> `a` ?
  * With this information we can find a plausible set of entries to return the users query for `vanam`.

I hope this explains the current state of affairs regarding normalization.

I think the hownorm1c normalization function `f`   (with modest refinements such as `am->a`) is currently satisfactory for searching.

However, I think we can practically start to develop refinements for the purpose of entry
alignment among dictionaries.  We should be able to parse good first approximations to grammatical
categories in many of the dictionaries, and to develop a refinement like `g` above.   Using this `g`
we can then get a better approximation to the entry equality function `EQ`.

**Comment by funderburkjim** [comment]
> cC -> C

There are about 4000 lines in  hwnorm1c with `cC` in spelling in some dictionary.

In 302 of these cases,  there are spellings without `cC` (i.e., only `C`) in some dictionaries, and `cC` in
  other dictionaries. [test_cC.txt](https://gist.github.com/funderburkjim/9653c5dd824ec27a1044ee92ad11fb31) has the list;  the spellings
are in SLP1.

So I guess I had noticed a few of these and that prompted me to include the rule.  Without the `cC->C` normalization, these 300 case would be split into separate entries in hwnorm1c.txt,  while the other
3700 cC spellings would have their normalizations changed to cC, but would otherwise be unchanged.

Do others think the 'cC'->'C' rule should be 
* removed ?
* kept as it is ?
* refined in some way suggested by the 300 examples?

Example of how to read a line of test_cC.txt:
```
EXAMPLE 1
akzaraCandas:akzaracCandas:CAE,CCS,GST,PD,PW,PWG,VCP;akzaraCandas:MW

akzaraCandas = normalized spelling
akzaracCandas = spelling in dictionaries CAE,CCS,GST,PD,PW,PWG,VCP
akzaraCandas = spelling in dictionary MW

EXAMPLE 2
aCandoma:acCandoma:PD;aCandoma:PD

aCandoma = normalized spelling
acCandoma:PD   spelling is acCandoma in dictionary(ies) PD
aCandoma:PD  spelling is aCandoma in dictionary PD  (SO PD HAS BOTH SPELLINGS!)
```

**Comment by funderburkjim** [comment]
#### all_cC.txt
This file has been added to [the gist](https://gist.github.com/funderburkjim/9653c5dd824ec27a1044ee92ad11fb31).
It contains all 3998 lines in hwnorm1c.txt that contain 'cC'.

It might be useful in reviewing the propriety or not of the `cC -> C` normalization rule.

**Comment by funderburkjim** [comment]
#### am -> a rule?

In the course of the explanation of 'vana' above, it was noted that while there is an `ending aM -> a` rule
in the current normalization procedure, there is no `ending am -> a` rule.   
Should there be such a rule?
* For the rule:  spelling of nouns like `vanam` will be identified with the stem form `vana`.
* Against the rule:  verbs like `gam` would be given normalized form `ga`  (Yuck!).

The `gam` -> `ga` type cases are so egregious that probably we should NOT implement a universal `am->a`
rule.

Agree?

**Comment by SergeA** [comment]
I've analyzed the ch/cch list.
I propose to make the general normalization ch>cch (not cch>ch). See below.

Pāṇini's sūtras are not very clear for the case of ch/cch. But, as I understand the common interpretation, in brief: after a vowel **ch** is changed to **cch**, but after the long at the end of a pada this change is optional. E.g.: necesarily in **gacchati ; hrīcchati ; śiva-cchāyā** &c. Optionally in **lakṣmī-cchāyā / lakṣmī-chāyā** &c. By this rule **cch** after vowel is either necessary or possible. While **ch** after vowel in many (in most) cases is wrong. So changing cch>ch is undesirable.

Also exist other cases where **cch** is the result of sandhied **t/d+ch/ś**. E.g. ud+śvas = ucchvas &c. The change cch>ch will give for them wrong spellings, and so is undesirable.

In the ch_vs._cch list, there are several cases where **ch** stands at the end of a root, as 'prach' &c. This way those are written by MW and in dhatupathas. It would be better to normalize the final variant **cch** to **ch**, e.g. pracch>prach. But there is a problem here, as in some dics roots are written with anubandhas. E.g. WIL gives 'praccha' for 'pracch'. As in this case **cch** is not final, and it will be too complicated to make a good rule for normalization. So I think there will not be much harm if these root-final cases will be also normalized with a common rule of ch>cch (after vowel). E.g. 'prach' will become 'pracch'.

There is also another group where the **cch** is the result of the duplication **after 'r'**. E.g. mūrcchat (<mūrchat). In these cases rcch>rch must be normalized by the corresponding rule.

There are few cases of duplicated **cch** after consonants other than 'r'.
ayukcchada in BUR,SCH,SHS,WIL,YAT  (against non-duplicated form in MD,MW,PW,PWG,VCP,SKD)
piñccha in AP90 
chancchan in WIL
śiloñccha in BUR
All have non-duplicated alternatives in other dics.
I consider these rare cases to be errors and propose to normalize them (after consonant) cch>ch.
(Ayuj+chada = ayuk-chada with no reason for duplication. Other are too strange by themselves.)

**So I propose to make 2 normalization rules:
(after vowel) ch>cch
(after consonant) cch>c**h

Also PGN gives 'cchandaka' (with 'chandaka' var.) with first duplicated **cch**, which also looks like nonsense. As this is a marginal case, I'm in doubt about existence of other examples, and I don´t think it requires a special rule for the normalization cch>ch in the beginning of the word.
______________________
Some additional observation on MW.
* The first **ch** is doubled after prefixes: chid >> vi-cchid ; ā-cchid
The spelling anāchṛṇṇa is corrected in p. 1311 > anācchṛṇṇa (from an+ā+chṛṇṇa, where duplication after ā is obligatory).
* But in compounds with first member ending long vowel, MW tends to keep undoubled:
antaḥ-sthā-chandas ; mahāchada ; mahāchāya ; mahāchidrā
(? also jihvā... -chedana ; nāsā... -chinnī )
But in some cases he gives two readings:
kathā-chala or -cchala ; kanyakā-chala or -cchala
* MW´s spelling  'reṣma... -chinna' (against PW,PWG´s reṣma-cchinna) where the duplication after short vowel is expected, can be considered print error, or we can say MW left to compose the sandhi to the reader (but for ch/cch in MW it is not usual), or perhaps it was an exact quoting of Vedic spelling.
* MW´s spelling 'avāchid' (repeated later in AP57!) must be considered a print error, as it is derived from ācchid, which is given correctly.
* An interesting case of MW´s 'prarcchaka' & 'prārcchaka' is an example where **cch** is duplicated not because it stands after 'r', but due to derivation from 'ṛcchaka' where 'ch' is obligatory duplicated after a short vowel.

**Comment by SergeA** [comment]
> The gam -> ga type cases are so egregious that probably we should NOT implement a universal am->a
> rule.

So gam, kram. nam, śam, yam &c. will turn to ga, kra, na, śa, ya &c.? Yeah, it´s a quite painful side effect.

**Comment by funderburkjim** [comment]
### More data on C, cC
@SergeA   I've done a test run of the normalization rule you suggest, and prepared some displays for 
review.  

To repeat, the test rule has two parts, expressed in SLP1 spelling as:
* vowel + C -> vowel + cC
* consonant + cC -> consonant + C

After reviewing the summary displays, I think that these rules are an improvement on the current
global `cC -> C` rule, and should replace those rules in the spelling normalization algorithm.

This [C_study gist](https://gist.github.com/funderburkjim/2386d74fca1036ee0e2680a20658e06a) has the summary files.

One strong argument in favor of the adoption of the revised rules is that the rules are in fact followed by most spellings in the various dictionaries.  Of the 6273 different spellings examined,  5691 (= 4891 + 800) or 91% are in accordance with the rules (i.e., vowel+cC, consonant+C).

After allowing a day or two for further comments, I'll go ahead and change the C normalization rule to the new form.

**Comment by funderburkjim** [comment]
> MW avāchid

MW72 shows avācchid, additional evidence in favor of considering avāchid in MW a print error.

**Comment by funderburkjim** [comment]
#### new C rule installed.
Per discussion above

#### avACid print change
Submitted corrections to Correction Form for change avACid -> avAcCid  in dictionaries MW, AP(1957).

When these corrections are installed, I'll add a 'print-change' record for each.  
@SergeA    In the print change record, one field is the name of the person responsible.  What name
should I put for you?

**Comment by funderburkjim** [comment]
#### prarcCaka
Have not changed normalization rule for this case, since it a rare side case, so normalized to prarCaka.
Similar for prArcCaka.

Noticed that PW, PWG have the prarCaka, prArCaka spellings -- is this a case where MW corrected PW, rather than copied?

Also noticed MW quotes the verb as prArC.

**Comment by gasyoun** [comment]
> This refinement is not possible with the current implementation of the normalization protocol.

So the only way to fix the adv. issue - whitelisting.

> NOTE: we don't find vanam (so we miss the entry in AP90). Maybe we should add rule am -> a ?

At least in the `vanam` case it makes sense.

> verbs like gam would be given normalized form ga (Yuck!).

Whitelist:
kam
gam
cam
ścam
jam
tam
dam
dham
nam
yam
ram
kram
dram
śram
lam
klam
vam
śam
kṣam
bhām

> So gam, kram. nam, śam, yam &c. will turn to ga, kra, na, śa, ya &c.? Yeah, it´s a quite painful side effect.

No need to, see above list from Whitney.

> I think the hownorm1c normalization function f (with modest refinements such as am->a) is currently satisfactory for searching.

Agree. The `g` would require months to develop.

> we can practically start to develop refinements for the purpose of entry
alignment among dictionaries. 

Sure, a test drive.

> certainly will corrupt many spellings like gacchat

As we speek of headwords and nothing else, so not such a big issue. There are no `gacchat` headowords. But agree with Serge on `changing cch>ch is undesirable`.

> So I think there will not be much harm if these root-final cases will be also normalized with a common rule of ch>cch (after vowel). E.g. 'prach' will become 'pracch'.

I'm against `pracch`. There are allready millions of dhatu forms, no need to make more. Whitelisting will help us out:

ich
yach
prach
mṛch
mlech
uñch
rañch
lāñch
mūrch
hūrch

> perhaps it was an exact quoting of Vedic spelling

•reṣmachinnaṃ yathā tṛṇma # AVŚ.6.102.2c.

as per https://yadi.sk/d/KWrfIDRF3P2pWD
Maurice Bloomfield's A Vedic Concordance

> Of the 6273 different spellings examined, 5691 (= 4891 + 800) or 91% are in accordance with the rules (i.e., vowel+cC, consonant+C).

Strong argument, agree.

**Comment by SergeA** [comment]
> @SergeA In the print change record, one field is the name of the person responsible. What name
> should I put for you?

Isn´t "SergeA" ok for this?

> Noticed that PW, PWG have the prarCaka, prArCaka spellings -- is this a case where MW corrected PW, rather than copied?

It seems so.

> I'm against pracch. There are allready millions of dhatu forms, no need to make more. Whitelisting will help us out

I'm too not very happy with pracch. But we do not invent here no new form, as pracch already exists in the dics. Please tell, if you know how to solve to divergence of the same root given by dics as pṛch-prach-pracch-praccha. The latter case ending in -a is an obstacle for white-listing by the last letter. Perhaps we need a separate normalization for SKD and others who spell roots with -a.

**Comment by gasyoun** [comment]
> Perhaps we need a separate normalization for SKD and others who spell roots with -a.

Maybe, what's your take @drdhaval2785 ?

**Comment by funderburkjim** [comment]
> Isn´t "SergeA" ok for this?

I was thinking that your real name might be preferable to your GitHub user name,  but I can use the GitHub user name if you prefer.

**Comment by funderburkjim** [comment]
> if you know how to solve to divergence of the same root given by dics as pṛch-prach-pracch-praccha. 

The answer is 'only incompletely' .   I think that we are close to the stage where this kind of 
correlation among root-spellings among dictionaries can be met head on.  We have made previous
attempts (such as with Wilson and MW in [WIL repository](https://github.com/sanskrit-lexicon/WIL)), but they must be viewed as preliminary.   Incidentally, Wilson spells roots with anubandhas, many with '-a' as in skd.

I've also made a
good attempt at correlating MW roots with Dhatupatha roots of Madhaviya via the Westergaard intermediary  (which is MW's Dhatupatha), but am not sure at this time how this work relates to the correlation of root spellings among the dictionaries.

**Comment by gasyoun** [comment]
> not sure at this time how this work relates to the correlation of root spellings among the dictionaries

Very well. Your preliminary work is not preliminary at all, it meets all expectations.

**Comment by funderburkjim** [comment]
> preliminary

It is preliminary in part because it has been applied to only two dictionaries: MW and WIL.

To extend it, I would probably think of correlating the roots in WIL, SHS, and YAT, since the latter two are so closely related to WIL.

Then I would correlate AP90 to MW, and then AP.

PW with PWG.

Then, maybe, BUR and STC.

Then SKD, VCP.
Then, aggregate all the correlations.

These comments are very sketchy, surely.  But having a good and accurate correlation of the root entries of the major dictionaries does seem like a worthy and achievable goal.

**Comment by gasyoun** [comment]
> But having a good and accurate correlation of the root entries of the major dictionaries does seem like a worthy and achievable goal

That's what supplement to my PhD thesis was all about https://yadi.sk/i/UpxRU_YgcmQMs

**Comment by funderburkjim** [comment]
That table should prove useful when we do  a root concordance.

I've also done an MW-WhitneyRoots concordance and MW-Westergaard Dhatupatha, which are basis of
the links:
![image](https://user-images.githubusercontent.com/6393033/32349650-2d9e068c-bfee-11e7-8292-2dea5216b02e.png)
This work should also be relevant.
